### PR TITLE
refactor: Refactor starts_with and ends_with for string

### PR DESCRIPTION
**Pull request by reswqa**

### CODE REVIEW

**Comment by stinodego** [comment]
Hey @reswqa , just a small tip: you can open a PR as a draft, see docs here:
https://github.blog/2019-02-14-introducing-draft-pull-requests/

Then we won't get a notification until you mark it as ready :)

File: crates/polars-plan/src/dsl/function_expr/strings.rs [hunk 1]
```diff
@@ -260,46 +261,36 @@ pub(super) fn ends_with(s: &[Series]) -> PolarsResult<Series> {
...
-            .map(|(opt_src, opt_val)| match (opt_src, opt_val) {
-                (Some(src), Some(val)) => src.ends_with(val),
-                _ => false,
-            })
-            .collect_trusted(),
-    };
-
-    out.rename(ca.name());
-    Ok(out.into_series())
+        _ => binary_elementwise(ca, sub, |opt_src, opt_sub| match (opt_src, opt_sub) {
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-plan/src/dsl/function_expr/strings.rs:269]: Indeed. That's faster.

**Review by ritchie46 — APPROVED** [review]

File: crates/polars-plan/src/dsl/function_expr/strings.rs [hunk 2]
```diff
@@ -260,46 +261,36 @@ pub(super) fn ends_with(s: &[Series]) -> PolarsResult<Series> {
     let ca = s[0].utf8()?;
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-plan/src/dsl/function_expr/strings.rs:261]: That's better indeed. Saves some code compilation.

File: crates/polars-plan/src/dsl/function_expr/strings.rs [hunk 3]
```diff
@@ -258,39 +258,15 @@ pub(super) fn contains(s: &[Series], literal: bool, strict: bool) -> PolarsResul
...
-            None => BooleanChunked::full(ca.name(), false, ca.len()),
-        },
-        _ => binary_elementwise(ca, sub, |opt_src, opt_sub| match (opt_src, opt_sub) {
-            (Some(src), Some(sub)) => Some(src.starts_with(sub)),
-            _ => Some(false),
-        }),
-    }
-    .with_name(ca.name())
-    .into_series())
+    let ca = s[0].cast(&DataType::Binary)?;
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/function_expr/strings.rs:266]: We have `as_binary` for that on `Utf8Chunked`.

File: crates/polars-ops/src/chunked_array/binary/namespace.rs [hunk 4]
```diff
@@ -46,6 +47,38 @@ pub trait BinaryNameSpaceImpl: AsBinary {
         out
     }
 
+    fn starts_with_chunked<B: AsBinary>(&self, prefix: &B) -> BooleanChunked {
```
[the remark this answers is not in the document]
Comment (ritchie46) [reply crates/polars-ops/src/chunked_array/binary/namespace.rs:50]: Indeed. `polars-plan` and `polars-lazy` should ideally only dispatch to the operations implemented by other crates.

Comment (ritchie46) [inline crates/polars-ops/src/chunked_array/binary/namespace.rs:50]: I don't think we need this to be generic right? 

As in, if we make this generic, this code will get compiled for every unique instantiation which leads to more binary bloat. I don't think in this case the bloat is worth it as we can always just convert to `BinaryChunked` before calling this function.

**Pull request merged**